neuro_data/movies/stats.py

```diff
@@ -146,3 +149,261 @@ def _make_tuples(self, key):
+            inputs[i] = dataset[index].inputs
+        assert np.all(np.abs(np.diff(inputs, axis=0)) == 0), 'Images of oracle trials do not match'
+
+    def sample_frames_from_dataset(self, target_indices, dataset, min_frames, num_of_neurons):
+        # Compute start_index
+        starting_index = 0
+        
+        upper_bound = np.min([dataset[index].responses.shape[0] for index in target_indices])
+        
+        if upper_bound - min_frames != 0:
+            starting_index = np.random.randint(0, upper_bound)
+
+        response_matrix = np.empty(shape=[target_indices.size, min_frames, num_of_neurons])
+        for i, index in enumerate(target_indices):
+            response_matrix[i] = dataset[index].responses[starting_index:starting_index+min_frames]
+
+        return response_matrix
+
+    def compute_oracle(self, outputs):
+        r = outputs.shape[0]
+        mu = outputs.mean(axis=0, keepdims=True)
+        oracles = ((mu - outputs / r) * r / (r - 1)).reshape(-1, outputs.shape[-1])
+        return oracles
+
+    def sample_and_compute_oracle(self, dataset, condition_hashes, sample_size, min_frames):
+        num_of_neurons = dataset[0].responses.shape[1]
+        dataset_condition_hashes = dataset.condition_hashes
+
+        # Oracle compuatation
+        true_responses = np.empty(shape=[len(condition_hashes), sample_size * min_frames, num_of_neurons])
+        true_oracles = np.empty(shape=[len(condition_hashes), sample_size * min_frames, num_of_neurons])
+
+        null_responses = np.empty(shape=[len(condition_hashes), sample_size * min_frames, num_of_neurons])
+        null_oracles = np.empty(shape=[len(condition_hashes), sample_size * min_frames, num_of_neurons])
+
+
+        for i in range(0, len(condition_hashes)):
+            # True Oracle Computation
+            # For each condition_hashes, sample (sample_size) trials to construct the true_response_matrix
+            # Select (sample_size) trials
+
+            true_target_indices = self.sample_from_condition_hash(condition_hashes[i], dataset_condition_hashes, sample_size)
+            
+            response_matrix = self.sample_frames_from_dataset(true_target_indices, dataset, min_frames, num_of_neurons)
+            self.check_input(true_target_indices, dataset)
+            true_responses[i] = response_matrix.reshape(-1, response_matrix.shape[-1])
+            true_oracles[i] = self.compute_oracle(response_matrix)
+
+            # Null Oracle Computation
+            # Select (samples_size) hashes and sample from them
+            target_hashes = np.random.choice(dataset_condition_hashes, sample_size, replace=False)
+
+            # Get target_indices
+            null_target_indices = np.array([self.sample_from_condition_hash(h, dataset_condition_hashes, 1)[0] for h in target_hashes])
+
+            # Sample for each target index
+            response_matrix = self.sample_frames_from_dataset(null_target_indices, dataset, min_frames, num_of_neurons)
```

Comment:
BUG: the null responses have all the same frame indices. You want this to be random.